l2geth/rollup/fees/rollup_fee.go

```diff
@@ -0,0 +1,108 @@
+package fees
+
+import (
+	"errors"
+	"fmt"
+	"math/big"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/params"
+)
+
+// overhead represents the fixed cost of batch submission of a single
+// transaction in gas
+const overhead uint64 = 4200
+
+// hundredMillion is a constant used in the gas encoding formula
+const hundredMillion uint64 = 100_000_000
+const hundredBillion uint64 = 100_000_000_000
+const FeeScalar uint64 = 1000
+
+var bigHundredMillion = new(big.Int).SetUint64(hundredMillion)
+var bigHundredBillion = new(big.Int).SetUint64(hundredBillion)
+var BigFeeScalar = new(big.Int).SetUint64(FeeScalar)
+
+// errInvalidGasPrice is the error returned when a user submits an incorrect gas
+// price. The gas price must satisfy a particular equation depending on if it
+// is a L1 gas price or a L2 gas price
+var errInvalidGasPrice = errors.New("rollup fee: invalid gas price")
+
+// CalculateFee calculates the fee that must be paid to the Rollup sequencer, taking into
+// account both the cost of submitting the transaction to L1 as well as
+// executing the transaction on L2
+// fee = (floor((l1GasLimit*l1GasPrice + l2GasLimit*l2GasPrice) / max(tx.gasPrice, 1)) + l2GasLimit) * tx.gasPrice
+// where tx.gasPrice is hard coded to 1000 * wei
```

Comment:
Maybe let's set the gasPrice to something larger, e.g. 1-100gwei, so that the gasLimit is not shown as a ridiculously large number to the user?

I think it'd be useful to add a doc explaining the bitpacking being done here:

```
// This function's return data encodes both the total fee which will be paid by the user and the
// gas limit which is expected to be used by the transaction
// [ totalFee | l2GasLimit ] 
```

Reply:
Setting the gasPrice that large will make the L2 gasPrice unrecoverable. There needs to be enough low order bits to fit ~15 million.

```golang
func DecodeL2GasLimit(gasLimit *big.Int) *big.Int {
	return new(big.Int).Mod(gasLimit, bigHundredMillion)
}
```